extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp

```diff
@@ -0,0 +1,213 @@
+ public:
+  SimpleSSLTestServerTLSv1_2(const std::string& port, const std::string& path)
+      : SimpleSSLTestServer(TLSv1_2_server_method(), port, path) {
+  }
+};
+
+class TLSClientSocketSupportedProtocolsTest {
+ public:
+  explicit TLSClientSocketSupportedProtocolsTest(const std::string& key_dir)
+      : key_dir_(key_dir), configuration_(std::make_shared<minifi::Configure>()) {
+  }
+
+  void run() {
+    configureSecurity();
+
+    verifyTLSClientSocketExclusiveCompatibilityWithTLSv1_2();
+  }
+
+
+ protected:
+  void configureSecurity() {
+    host_ = org::apache::nifi::minifi::io::Socket::getMyHostName();
+    port_ = "38777";
+    if (!key_dir_.empty()) {
+      configuration_->set(minifi::Configure::nifi_remote_input_secure, "true");
+      configuration_->set(minifi::Configure::nifi_security_client_certificate, key_dir_ + "cn.crt.pem");
+      configuration_->set(minifi::Configure::nifi_security_client_private_key, key_dir_ + "cn.ckey.pem");
+      configuration_->set(minifi::Configure::nifi_security_client_pass_phrase, key_dir_ + "cn.pass");
+      configuration_->set(minifi::Configure::nifi_security_client_ca_certificate, key_dir_ + "nifi-cert.pem");
+      configuration_->set(minifi::Configure::nifi_default_directory, key_dir_);
+    }
+  }
+
+  void verifyTLSClientSocketExclusiveCompatibilityWithTLSv1_2() {
+    verifyTLSProtocolIncompatibility<SimpleSSLTestServerTLSv1>();
+    verifyTLSProtocolIncompatibility<SimpleSSLTestServerTLSv1_1>();
+    verifyTLSProtocolCompatibility<SimpleSSLTestServerTLSv1_2>();
+  }
+
+  template <class TLSTestSever>
+  void verifyTLSProtocolIncompatibility() {
+    verifyTLSProtocolCompatibility<TLSTestSever>(false);
+  }
+
+  template <class TLSTestSever>
+  void verifyTLSProtocolCompatibility() {
+    verifyTLSProtocolCompatibility<TLSTestSever>(true);
+  }
+
+  template <class TLSTestSever>
+  void verifyTLSProtocolCompatibility(const bool should_be_compatible) {
+    TLSTestSever server(port_, key_dir_);
+    server.waitForConnection();
+
+    const auto socket_context = std::make_shared<org::apache::nifi::minifi::io::TLSContext>(configuration_);
+    client_socket_ = std::make_shared<org::apache::nifi::minifi::io::TLSSocket>(socket_context, host_, std::stoi(port_), 0);
+    const bool client_initialized_successfully = (client_socket_->initialize() == 0);
+    assert(client_initialized_successfully == should_be_compatible);
+    server.shutdownServer();
+    assert(server.hadConnection() == should_be_compatible);
+  }
+
+ protected:
+    std::shared_ptr<org::apache::nifi::minifi::io::TLSSocket> client_socket_;
+    std::string host_;
+    std::string port_;
+    std::string key_dir_;
+    std::shared_ptr<minifi::Configure> configuration_;
+};
+
+static void sigpipe_handle(int) {
+}
+
+int main(int argc, char **argv) {
+  std::string key_dir;
+  if (argc > 1) {
+    key_dir = argv[1];
+  }
+#ifndef WIN32
+  signal(SIGPIPE, sigpipe_handle);
```

Comment:
This overrides how to handle the broken pipe signal (which will happen because we are testing that you cannot connect with outdated protocol). The default behavior for this signal is to end the process.
This workaround is already used by extensions/standard-processors/tests/integration/SecureSocketGetTCPTest.cpp (I used that class as template for these tests)